File: bindings/java/src/test/java/io/omq/OptionsTest.java
package io.omq;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.Arrays;
import org.junit.jupiter.api.Test;

final class OptionsTest {
    @Test
    void rejectsNegativeHighWaterMarks() {
        try (Context context = OMQ.context();
             Socket push = context.socket(SocketType.PUSH)) {
            assertThrows(IllegalArgumentException.class, () -> push.sendHighWaterMark(-1));
            assertThrows(IllegalArgumentException.class, () -> push.receiveHighWaterMark(-1));
        }
    }

    @Test
    void rejectsNegativeDurations() {
        try (Context context = OMQ.context();
             Socket push = context.socket(SocketType.PUSH)) {
            assertThrows(IllegalArgumentException.class, () -> push.linger(Duration.ofMillis(-1)));
            assertThrows(IllegalArgumentException.class,
                    () -> push.heartbeatInterval(Duration.ofMillis(-1)));
            assertThrows(IllegalArgumentException.class,
                    () -> push.handshakeTimeout(Duration.ofMillis(-1)));
        }
    }

    @Test
    void acceptsPreMaterializationOptions() {
        try (Context context = OMQ.context();
             Socket push = context.socket(SocketType.PUSH)) {
            push.linger(Duration.ZERO)
                    .lingerForever()
                    .workloadProfile(WorkloadProfile.LATENCY)
                    .defaultWorkloadProfile()
                    .reconnectDisabled()
                    .reconnectInterval(Duration.ofMillis(100))
                    .reconnectExponential(Duration.ofMillis(10), Duration.ofSeconds(1))
                    .reconnectStopConnRefused(true)
                    .sendHighWaterMark(10)
                    .receiveHighWaterMark(11)
                    .heartbeatInterval(Duration.ofMillis(100))
                    .heartbeatTtl(Duration.ofSeconds(3))
                    .noHeartbeatTtl()
                    .heartbeatTimeout(Duration.ofSeconds(5))
                    .defaultHeartbeatTimeout()
                    .heartbeatOff()
                    .handshakeTimeout(Duration.ofSeconds(1))
                    .maxPendingHandshakes(8)
                    .noMaxMessageSize()
                    .conflate(false)
                    .routerMandatory(false)
                    .onMute(OnMute.BLOCK)
                    .onMute(OnMute.DROP_NEWEST)
                    .onMute(OnMute.DROP_OLDEST)
                    .tcpKeepalive(Duration.ofSeconds(60), Duration.ofSeconds(10), 3)
                    .tcpKeepaliveOff()
                    .tcpKeepaliveDefault()
                    .sendBufferSize(65_536)
                    .defaultSendBufferSize()
                    .receiveBufferSize(65_536)
                    .defaultReceiveBufferSize()
                    .compressionDict("dict".getBytes())
                    .noCompressionDict()
                    .compressionThreshold(128)
                    .compressionDefaultThreshold()
                    .compressionLevel(1)
                    .compressionDefaultLevel()
                    .compressionDictCapacity(2_048)
                    .defaultCompressionDictCapacity()
                    .maxReceiveDictSize(8_192)
                    .defaultMaxReceiveDictSize()
                    .compressionOffloadThreshold(8_192)
                    .noCompressionOffload()
                    .largeMessageThreshold(4_096)
                    .disableLargeMessagePath()
                    .arenaThreshold(65_536)
                    .defaultArenaThreshold()
                    .transmitSlotCapacity(2 * 1024 * 1024)
                    .defaultTransmitSlotCapacity()
                    .xpubNoDrop(false);
        }
    }

    @Test
    void builderAppliesReusableOptionsBeforeMaterialization() {
        SocketOptions options = SocketOptions.builder()
                .linger(Duration.ZERO)
                .lingerForever()
                .workloadProfile(WorkloadProfile.LATENCY)
                .defaultWorkloadProfile()
                .reconnectDisabled()
                .reconnectInterval(Duration.ofMillis(100))
                .reconnectExponential(Duration.ofMillis(10), Duration.ofSeconds(1))
                .reconnectStopConnRefused(true)
                .sendHighWaterMark(10)
                .receiveHighWaterMark(11)
                .heartbeatInterval(Duration.ofMillis(100))
                .heartbeatTtl(Duration.ofSeconds(3))
                .noHeartbeatTtl()
                .heartbeatTimeout(Duration.ofSeconds(5))
                .defaultHeartbeatTimeout()
                .heartbeatOff()
                .handshakeTimeout(Duration.ofSeconds(1))
                .maxPendingHandshakes(8)
                .noMaxMessageSize()
                .conflate(false)
                .routerMandatory(false)
                .onMute(OnMute.BLOCK)
                .onMute(OnMute.DROP_NEWEST)
                .onMute(OnMute.DROP_OLDEST)
                .tcpKeepalive(Duration.ofSeconds(60), Duration.ofSeconds(10), 3)
                .tcpKeepaliveOff()
                .tcpKeepaliveDefault()
                .sendBufferSize(65_536)
                .defaultSendBufferSize()
                .receiveBufferSize(65_536)
                .defaultReceiveBufferSize()
                .compressionDict("dict".getBytes(StandardCharsets.UTF_8))
                .noCompressionDict()
                .compressionAutoTrain(true)
                .compressionThreshold(128)
                .compressionDefaultThreshold()
                .compressionLevel(1)
                .compressionDefaultLevel()
                .compressionDictCapacity(2_048)
                .defaultCompressionDictCapacity()
                .maxReceiveDictSize(8_192)
                .defaultMaxReceiveDictSize()
                .compressionOffloadThreshold(8_192)
                .noCompressionOffload()
                .largeMessageThreshold(4_096)
                .disableLargeMessagePath()
                .arenaThreshold(65_536)
                .defaultArenaThreshold()
                .transmitSlotCapacity(2 * 1024 * 1024)
                .defaultTransmitSlotCapacity()
                .xpubNoDrop(false)
                .build();

        try (Context context = OMQ.context();
             Socket push = context.socket(SocketType.PUSH, options)) {
            assertEquals(Socket.class, push.getClass());
        }
    }

    @Test
    void builderCopiesByteArrayOptions() {
        byte[] identity = "before".getBytes(StandardCharsets.UTF_8);
        SocketOptions options = SocketOptions.builder().identity(identity).build();
        Arrays.fill(identity, (byte) 'x');

        try (Context context = OMQ.context();
             Socket router = context.socket(SocketType.ROUTER);
             Socket dealer = context.socket(SocketType.DEALER, options)) {
            String endpoint = router.bind(TestSupport.inprocEndpoint("builder-identity"));
            dealer.connect(endpoint);
            dealer.send("hello");

            Message request = router.receive(Duration.ofSeconds(5)).orElseThrow();
            assertEquals("before", new String(request.part(0), StandardCharsets.UTF_8));
        }
    }

    @Test
    void builderConfiguresPlainAuth() {
        SocketOptions serverOptions = SocketOptions.builder()
                .plainServer("alice", "secret")
                .build();
        SocketOptions clientOptions = SocketOptions.builder()
                .plainClient("alice", "secret")
                .build();

        try (Context context = OMQ.context();
             Socket pull = context.socket(SocketType.PULL, serverOptions);
             Socket push = context.socket(SocketType.PUSH, clientOptions)) {
            String endpoint = pull.bind("tcp://127.0.0.1:0");
            push.connect(endpoint);
            push.waitConnected(1, Duration.ofSeconds(5));
            push.send("hello over builder plain");

            assertEquals("hello over builder plain", pull.receive(Duration.ofSeconds(5)).orElseThrow().text());
        }
    }

    @Test
    void builderRejectsInvalidOptionsEarly() {
        assertThrows(IllegalArgumentException.class,
                () -> SocketOptions.builder().sendHighWaterMark(-1));
        assertThrows(IllegalArgumentException.class,
                () -> SocketOptions.builder().compressionDict(new byte[0]));
        assertThrows(IllegalArgumentException.class,
                () -> SocketOptions.builder().compressionLevel(5));
        assertThrows(IllegalArgumentException.class,
                () -> SocketOptions.builder().heartbeatTtl(Duration.ofMillis(6_553_501)));
    }

    @Test
    void optionsCannotChangeAfterConnect() {
        try (Context context = OMQ.context();
             Socket push = context.socket(SocketType.PUSH)) {
            push.connect("tcp://127.0.0.1:" + TestSupport.freePort());
            assertThrows(OMQException.class, () -> push.linger(Duration.ZERO));
            assertThrows(OMQException.class, () -> push.routerMandatory(true));
        }
    }

    @Test
    void rejectsNegativeMaxMessageSize() {
        try (Context context = OMQ.context();
             Socket pull = context.socket(SocketType.PULL)) {
            assertThrows(IllegalArgumentException.class, () -> pull.maxMessageSize(-1));
        }
    }

    @Test
    void rejectsNegativeCompressionThreshold() {
        try (Context context = OMQ.context();
             Socket pull = context.socket(SocketType.PULL)) {
            assertThrows(IllegalArgumentException.class, () -> pull.compressionThreshold(-1));
        }
    }

    @Test
    void rejectsInvalidExtendedOptions() {
        try (Context context = OMQ.context();
             Socket pull = context.socket(SocketType.PULL)) {
            assertThrows(IllegalArgumentException.class,
                    () -> pull.reconnectExponential(Duration.ofSeconds(2), Duration.ofSeconds(1)));
            assertThrows(IllegalArgumentException.class, () -> pull.maxPendingHandshakes(0));
            assertThrows(IllegalArgumentException.class,
                    () -> pull.tcpKeepalive(Duration.ofSeconds(1), Duration.ofSeconds(1), 0));
            assertThrows(IllegalArgumentException.class, () -> pull.sendBufferSize(-1));
            assertThrows(IllegalArgumentException.class, () -> pull.receiveBufferSize(-1));
            assertThrows(IllegalArgumentException.class, () -> pull.compressionDictCapacity(-1));
            assertThrows(IllegalArgumentException.class, () -> pull.maxReceiveDictSize(-1));
            assertThrows(IllegalArgumentException.class,
                    () -> pull.compressionOffloadThreshold(-1));
            assertThrows(IllegalArgumentException.class, () -> pull.largeMessageThreshold(-1));
            assertThrows(IllegalArgumentException.class, () -> pull.arenaThreshold(-1));
            assertThrows(IllegalArgumentException.class, () -> pull.transmitSlotCapacity(-1));
        }
    }

    @Test
    void validatesOptionLimitsBeforeNativeCall() {
        try (Context context = OMQ.context();
             Socket pull = context.socket(SocketType.PULL)) {
            assertThrows(IllegalArgumentException.class, () -> pull.compressionDict(new byte[0]));
            assertThrows(IllegalArgumentException.class, () -> pull.compressionDict(new byte[9_000]));
            assertThrows(IllegalArgumentException.class, () -> pull.compressionLevel(-9));
            assertThrows(IllegalArgumentException.class, () -> pull.compressionLevel(5));
            assertThrows(IllegalArgumentException.class,
                    () -> pull.heartbeatTtl(Duration.ofMillis(6_553_501)));
        }
    }

    @Test
    void positiveSubMillisecondDurationsRoundUp() {
        assertEquals(0, Socket.millis(Duration.ZERO));
        assertEquals(1, Socket.millis(Duration.ofNanos(1)));
        assertEquals(2, Socket.millis(Duration.ofNanos(1_000_001)));
    }

    @Test
    void hugeDurationsSaturate() {
        assertEquals(Long.MAX_VALUE, Socket.millis(Duration.ofSeconds(Long.MAX_VALUE)));
    }
}

File: bindings/go/options_test.go
package omq

import (
	"errors"
	"testing"
	"time"
)

func TestOptionsAcceptBeforeMaterialization(t *testing.T) {
	ctx := openTestContext(t)
	defer closeContext(t, ctx)

	push := newTestSocket(t, ctx, Push)
	defer closeSocket(t, push)

	options := []SocketOption{
		Linger(0),
		LingerForever(),
		Workload(WorkloadLatency),
		DefaultWorkload(),
		ReconnectDisabled(),
		ReconnectInterval(100 * time.Millisecond),
		ReconnectExponential(10*time.Millisecond, time.Second),
		ReconnectStopConnRefused(true),
		SendHWM(10),
		RecvHWM(11),
		HeartbeatInterval(100 * time.Millisecond),
		HeartbeatTTL(3 * time.Second),
		NoHeartbeatTTL(),
		HeartbeatTimeout(5 * time.Second),
		DefaultHeartbeatTimeout(),
		HeartbeatOff(),
		HandshakeTimeout(time.Second),
		MaxPendingHandshakes(8),
		NoMaxMessageSize(),
		Conflate(false),
		RouterMandatory(false),
		OnMutePolicy(OnMuteBlock),
		OnMutePolicy(OnMuteDropNewest),
		OnMutePolicy(OnMuteDropOldest),
		TCPKeepalive(time.Minute, 10*time.Second, 3),
		TCPKeepaliveOff(),
		TCPKeepaliveDefault(),
		SendBufferSize(65_536),
		DefaultSendBufferSize(),
		RecvBufferSize(65_536),
		DefaultRecvBufferSize(),
		CompressionDict([]byte("dict")),
		NoCompressionDict(),
		CompressionAutoTrain(true),
		CompressionThreshold(128),
		CompressionDefaultThreshold(),
		CompressionLevel(1),
		CompressionDefaultLevel(),
		CompressionDictCapacity(2_048),
		DefaultCompressionDictCapacity(),
		MaxRecvDictSize(8_192),
		DefaultMaxRecvDictSize(),
		CompressionOffloadThreshold(8_192),
		NoCompressionOffload(),
		LargeMessageThreshold(4_096),
		DisableLargeMessagePath(),
		ArenaThreshold(65_536),
		DefaultArenaThreshold(),
		TransmitSlotCapacity(2 * 1024 * 1024),
		DefaultTransmitSlotCapacity(),
		XPubNoDrop(false),
	}
	for _, option := range options {
		if err := option(push); err != nil {
			t.Fatal(err)
		}
	}
}

func TestOptionsCopyByteSlices(t *testing.T) {
	ctx := openTestContext(t)
	defer closeContext(t, ctx)

	identity := []byte("before")
	identityOption := Identity(identity)
	identity[0] = 'x'
	dealer, err := ctx.Socket(Dealer, identityOption)
	if err != nil {
		t.Fatal(err)
	}
	defer closeSocket(t, dealer)
	router := newTestSocket(t, ctx, Router)
	defer closeSocket(t, router)

	endpoint, err := router.Bind("inproc://go-options-copy-identity")
	if err != nil {
		t.Fatal(err)
	}
	if err := dealer.Connect(endpoint); err != nil {
		t.Fatal(err)
	}
	if err := dealer.SendTimeout(String("hello"), time.Second); err != nil {
		t.Fatal(err)
	}
	request, err := router.RecvTimeout(time.Second)
	if err != nil {
		t.Fatal(err)
	}
	if got := string(request.Part(0)); got != "before" {
		t.Fatalf("identity = %q, want before", got)
	}

	options := dealer.Options()
	if !options.Identity.Set || string(options.Identity.Value) != "before" {
		t.Fatalf("identity option = %q/%v, want before/true", options.Identity.Value, options.Identity.Set)
	}
	options.Identity.Value[0] = 'z'
	again := dealer.Options()
	if string(again.Identity.Value) != "before" {
		t.Fatalf("identity option mutated = %q, want before", again.Identity.Value)
	}

	dict := []byte("abcd")
	dictOption := CompressionDict(dict)
	dict[0] = 'z'
	push, err := ctx.Socket(Push, dictOption)
	if err != nil {
		t.Fatal(err)
	}
	defer closeSocket(t, push)
	pushOptions := push.Options()
	if !pushOptions.CompressionDict.Set || string(pushOptions.CompressionDict.Value) != "abcd" {
		t.Fatalf("compression dict option = %q/%v, want abcd/true", pushOptions.CompressionDict.Value, pushOptions.CompressionDict.Set)
	}
	pushOptions.CompressionDict.Value[0] = 'y'
	pushAgain := push.Options()
	if string(pushAgain.CompressionDict.Value) != "abcd" {
		t.Fatalf("compression dict option mutated = %q, want abcd", pushAgain.CompressionDict.Value)
	}
}

func TestOptionsSnapshotReportsConfiguredValues(t *testing.T) {
	ctx := openTestContext(t)
	defer closeContext(t, ctx)

	socket, err := ctx.Socket(Push,
		SendHWM(7),
		RecvHWM(8),
		Linger(2*time.Millisecond),
		HeartbeatOff(),
		NoHeartbeatTTL(),
		DefaultHeartbeatTimeout(),
		NoHandshakeTimeout(),
		NoMaxMessageSize(),
		PlainClient("alice", "secret"),
		Workload(WorkloadLatency),
		ReconnectExponential(time.Millisecond, 10*time.Millisecond),
		ReconnectStopConnRefused(true),
		MaxPendingHandshakes(4),
		Conflate(true),
		RouterMandatory(true),
		OnMutePolicy(OnMuteDropNewest),
		TCPKeepalive(time.Second, time.Second, 2),
		SendBufferSize(1024),
		RecvBufferSize(2048),
		XPubNoDrop(true),
		CompressionAutoTrain(true),
		CompressionThreshold(64),
		CompressionLevel(1),
		CompressionDict([]byte("abcd")),
		CompressionDictCapacity(4096),
		MaxRecvDictSize(8192),
		CompressionOffloadThreshold(16384),
		LargeMessageThreshold(32768),
		ArenaThreshold(65536),
		TransmitSlotCapacity(131072),
	)
	if err != nil {
		t.Fatal(err)
	}
	defer closeSocket(t, socket)

	options := socket.Options()
	if !options.SendHWM.Set || options.SendHWM.Value != 7 {
		t.Fatalf("SendHWM = %#v", options.SendHWM)
	}
	if !options.RecvHWM.Set || options.RecvHWM.Value != 8 {
		t.Fatalf("RecvHWM = %#v", options.RecvHWM)
	}
	if !options.Linger.Set || options.Linger.Value != 2*time.Millisecond {
		t.Fatalf("Linger = %#v", options.Linger)
	}
	if !options.HeartbeatOff || !options.NoHeartbeatTTL || !options.DefaultHeartbeatTimeout ||
		!options.NoHandshakeTimeout || !options.NoMaxMessageSize {
		t.Fatalf("duration sentinel options = %#v", options)
	}
	if !options.PlainClient.Set || options.PlainClient.Value.Username != "alice" ||
		options.PlainClient.Value.Password != "secret" {
		t.Fatalf("PlainClient = %#v", options.PlainClient)
	}
	if !options.Workload.Set || options.Workload.Value != WorkloadLatency {
		t.Fatalf("Workload = %#v", options.Workload)
	}
	if !options.Reconnect.Set || options.Reconnect.Value.Mode != "exponential" {
		t.Fatalf("Reconnect = %#v", options.Reconnect)
	}
	if !options.ReconnectStopConnRefused.Set || !options.ReconnectStopConnRefused.Value {
		t.Fatalf("ReconnectStopConnRefused = %#v", options.ReconnectStopConnRefused)
	}
	if !options.MaxPendingHandshakes.Set || options.MaxPendingHandshakes.Value != 4 {
		t.Fatalf("MaxPendingHandshakes = %#v", options.MaxPendingHandshakes)
	}
	if !options.Conflate.Set || !options.Conflate.Value ||
		!options.RouterMandatory.Set || !options.RouterMandatory.Value ||
		!options.XPubNoDrop.Set || !options.XPubNoDrop.Value {
		t.Fatalf("bool options = %#v", options)
	}
	if !options.OnMute.Set || options.OnMute.Value != OnMuteDropNewest {
		t.Fatalf("OnMute = %#v", options.OnMute)
	}
	if !options.TCPKeepalive.Set || options.TCPKeepalive.Value.Mode != "enabled" ||
		options.TCPKeepalive.Value.Count != 2 {
		t.Fatalf("TCPKeepalive = %#v", options.TCPKeepalive)
	}
	if !options.SendBufferSize.Set || options.SendBufferSize.Value != 1024 ||
		!options.RecvBufferSize.Set || options.RecvBufferSize.Value != 2048 {
		t.Fatalf("buffer sizes = %#v/%#v", options.SendBufferSize, options.RecvBufferSize)
	}
	if !options.CompressionAutoTrain.Set || !options.CompressionAutoTrain.Value ||
		!options.CompressionThreshold.Set || options.CompressionThreshold.Value != 64 ||
		!options.CompressionLevel.Set || options.CompressionLevel.Value != 1 ||
		!options.CompressionDict.Set || string(options.CompressionDict.Value) != "abcd" {
		t.Fatalf("compression basics = %#v", options)
	}
	if !options.CompressionDictCapacity.Set || options.CompressionDictCapacity.Value != 4096 ||
		!options.MaxRecvDictSize.Set || options.MaxRecvDictSize.Value != 8192 ||
		!options.CompressionOffloadThreshold.Set || options.CompressionOffloadThreshold.Value != 16384 ||
		!options.LargeMessageThreshold.Set || options.LargeMessageThreshold.Value != 32768 ||
		!options.ArenaThreshold.Set || options.ArenaThreshold.Value != 65536 ||
		!options.TransmitSlotCapacity.Set || options.TransmitSlotCapacity.Value != 131072 {
		t.Fatalf("compression/perf sizes = %#v", options)
	}
}

func TestOptionsCannotChangeAfterMaterialization(t *testing.T) {
	ctx := openTestContext(t)
	defer closeContext(t, ctx)

	push := newTestSocket(t, ctx, Push)
	defer closeSocket(t, push)
	if err := push.Connect("tcp://127.0.0.1:1"); err != nil {
		t.Fatal(err)
	}
	if err := Linger(0)(push); !isConfigError(err) {
		t.Fatalf("Linger after connect err = %v, want ConfigError", err)
	}
	if err := RouterMandatory(true)(push); !isConfigError(err) {
		t.Fatalf("RouterMandatory after connect err = %v, want ConfigError", err)
	}
}

func TestOptionsRejectInvalidValues(t *testing.T) {
	ctx := openTestContext(t)
	defer closeContext(t, ctx)

	pull := newTestSocket(t, ctx, Pull)
	defer closeSocket(t, pull)

	invalid := []SocketOption{
		Linger(-time.Millisecond),
		HeartbeatInterval(-time.Millisecond),
		HeartbeatTTL(time.Duration(maxHeartbeatTTLMillis+1) * time.Millisecond),
		HandshakeTimeout(-time.Millisecond),
		MaxMessageSize(-1),
		ReconnectExponential(time.Second, time.Millisecond),
		MaxPendingHandshakes(0),
		TCPKeepalive(time.Second, time.Second, 0),
		SendBufferSize(-1),
		RecvBufferSize(-1),
		CompressionDict(nil),
		CompressionDict(make([]byte, compressionDictMaxBytes+1)),
		CompressionLevel(zstdLevelMax + 1),
		CompressionThreshold(-1),
		CompressionDictCapacity(-1),
		MaxRecvDictSize(-1),
		CompressionOffloadThreshold(-1),
		LargeMessageThreshold(-1),
		ArenaThreshold(-1),
		TransmitSlotCapacity(-1),
		Identity(make([]byte, zmtpMaxShortStringBytes+1)),
		PlainServerCredentials(PlainCredential{
			Username: string(make([]byte, zmtpMaxShortStringBytes+1)),
			Password: "secret",
		}),
		PlainServerCredentials(PlainCredential{Username: "has space", Password: "secret"}),
		PlainClient("alice", string(make([]byte, zmtpMaxShortStringBytes+1))),
		PlainServerAuth(nil),
		CurveServerAuth(CurveKeypair{}, nil),
	}
	for _, option := range invalid {
		if err := option(pull); !isConfigError(err) {
			t.Fatalf("option err = %v, want ConfigError", err)
		}
	}

	first, err := GenerateCurveKeypair()
	if err != nil {
		t.Fatal(err)
	}
	second, err := GenerateCurveKeypair()
	if err != nil {
		t.Fatal(err)
	}
	if err := CurveServer(CurveKeypair{Public: first.Public, Secret: second.Secret})(pull); !isConfigError(err) {
		t.Fatalf("CurveServer mismatched keypair err = %v, want ConfigError", err)
	}
}

func isConfigError(err error) bool {
	var config *ConfigError
	return errors.As(err, &config)
}
